### PR TITLE
Enhance `yii\behaviors\AttributeBehavior`

### PR DESCRIPTION
**Pull request by Kolyunya**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #11923 |

`AttributeBehavior` can preserve current attribute values now.
Basic tests coverage is included.

### CODE REVIEW

File: framework/behaviors/AttributeBehavior.php
Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: this statement should be readable without the comment. Can you use a native function as a replacement for !empty?

File: tests/framework/behaviors/AttributeBehaviorTest.php
Comment (dynasource) [inline tests/framework/behaviors/AttributeBehaviorTest.php]: can you wrap the test methods into 1 test method with a dataprovider? This enables code-reuse and its easier to compare the data used for testing.

File: framework/behaviors/AttributeBehavior.php
Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: 1. this statement should be formulated in such way that a comment is not necessary.
2. can you use a native function instead of empty?
3. a check on $event->name is missing, like with $skipUpdateOnClean. This makes it confusing an less readable and error prone.

**Review by dynasource — CHANGES_REQUESTED** [review]
thanks for adding the first tests for this class

File: tests/framework/behaviors/AttributeBehaviorTest.php
Comment (Kolyunya) [inline tests/framework/behaviors/AttributeBehaviorTest.php]: Done.

File: framework/behaviors/AttributeBehavior.php
Comment (Kolyunya) [inline framework/behaviors/AttributeBehavior.php]: @dynasource 

> this statement should be formulated in such way that a comment is not necessary.

It's not necessary, removed it.

> can you use a native function instead of empty?

Refactored that sub-condition to the `$this->owner->$attribute !== ''`.

> a check on $event->name is missing, like with $skipUpdateOnClean. This makes it confusing an less readable and error prone.

I can't get it what is the reason for checking the event name while evaluating each attribute? Can you elaborate please?

Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: > can you use a native function instead of empty?
> Refactored that sub-condition to the $this->owner->$attribute !== ''.

the logic behind this should be made less complex. The current format is not self explaining enough and takes time to interpretate and has room for improvement.`skipUpdateOnFilled` in combination with a false comparison is less readable than a positive one with a variable name like `forceUpdateOnFilled`. Additionally, `onFilled` is  a bit vague. What is filled in the context? When is it about a change or about being empty? It would expect something more related to your original issue description.

> a check on $event->name is missing, like with $skipUpdateOnClean. This makes it confusing an less readable and error prone.
see https://github.com/Kolyunya/yii2/blob/8cee4ec59b9553b47adad4cee7e3828ee08123cc/framework/behaviors/AttributeBehavior.php#L113. The current code looks to be run in the 'insert' context also which could be confusing.

Comment (Kolyunya) [inline framework/behaviors/AttributeBehavior.php]: @dynasource I used the `skipUpdateOn...` property because it is consistent with the `skipUpdateOnClean`. Do you really insist on naming the property inconsistently even if it serves the similar purpose? Maybe we can use `skipUpdateOnNonEmpty`?

If you insist on naming it differently I can suggest something like `preserveExisting`.

What property name would you suggest?

>  When is it about a change or about being empty?
It's about being empty.

Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: Well, I'd prefer clear naming above consistency. However, your suggestion of `skipUpdateOnNonEmpty` is a lot better.

Comment (Kolyunya) [inline framework/behaviors/AttributeBehavior.php]: @dynasource I refactored the code, what would you say about `preserveExistingValues`?

```php
if ($this->preserveExistingValues && !empty($this->owner->$attribute)) {
    continue;
}
$this->owner->$attribute = $value;
```

`empty` is used here since we want to skip the attribute evaluation both when it's an empty string and `null`.

No event name check is needed here since we want to skip non-empty attributes evaluation while insert too.

Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: its getting more readable. However, its still confusing with respect to the 2 scenario's 'insert' & 'update' as 'insert' is not having existingValues right?

Comment (Kolyunya) [inline framework/behaviors/AttributeBehavior.php]: >  'insert' is not having existingValues right?
Insert does have `existing` values if user provided them. User provides values (they exist since this time), then the behavior performs the checks and evaluates them if needed.

Comment (Kolyunya) [inline framework/behaviors/AttributeBehavior.php]: @dynasource
 >  'insert' is not having existingValues right?

Insert does have `existing` values if user provided them. User provides values (they exist since this time), then the behavior performs the checks and evaluates them if needed.

Comment (Kolyunya) [inline framework/behaviors/AttributeBehavior.php]: @dynasource do you think that `preserveNonEmptyValues` is better?

Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: thx, its getting better

Comment (Kolyunya) [inline framework/behaviors/AttributeBehavior.php]: @dynasource anything else we can improve?

Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: lets wait for the reviews from the rest of the team

**Review by samdark — APPROVED** [review]

**Review by dynasource — APPROVED** [review]

File: framework/behaviors/AttributeBehavior.php
Comment (klimov-paul) [inline framework/behaviors/AttributeBehavior.php]: Would `skipNotEmpty` be better?

Comment (Kolyunya) [inline framework/behaviors/AttributeBehavior.php]: @klimov-paul wouldn't `skipNonEmpty` be better?

Comment (samdark) [inline framework/behaviors/AttributeBehavior.php]: Yes. `Non` is better.

Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: @klimov-paul, @samdark, see https://github.com/yiisoft/yii2/pull/13586#discussion_r101737303 for the motivation for $preserveNonEmptyValues

Comment (samdark) [inline framework/behaviors/AttributeBehavior.php]: Actually yes, `preserveNonEmptyValues` is more clear despite being a bit longer.

Comment (SilverFire) [inline framework/behaviors/AttributeBehavior.php]: From my POV, `preserve` is a bit more complex word than `ommit` or `skip`

Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: skip + not is a double negation which is not ideal. Positive formulation is better for readability. What about `keepSetValues`

Comment (samdark) [inline framework/behaviors/AttributeBehavior.php]: Variable may be set to empty value so it's not precise this way.

Comment (dynasource) [inline framework/behaviors/AttributeBehavior.php]: `keepNonEmptyValues`

**Comment by cebe** [comment]
Note that there is `$immutable` property in subclass `SluggableBehavior` which does the same.

**Review by cebe — CHANGES_REQUESTED** [review]
Note that there is `$immutable` property in subclass `SluggableBehavior` which does the same.

**Comment by samdark** [comment]
@cebe Yes. It should be deprecated.

**Comment by cebe** [comment]
how about moving `$immutable` one layer up?

**Comment by Kolyunya** [comment]
@cebe semantics of `$immutable` means that it preserves values regardless of their emptiness. This may be misleading.

**Comment by samdark** [comment]
@cebe I agree with @Kolyunya.

**Comment by samdark** [comment]
Related to #13449

**Comment by Kolyunya** [comment]
Is here anything else I should fix of it was rescheduled to `2.0.13` because we weren't able to come up with an appropriate variable name yet?

**Comment by samdark** [comment]
Merged with minor adjustments https://github.com/yiisoft/yii2/commit/9031a115b9c85d9656e516ca30854d8ec969942d, thank you!

**Pull request closed without merging**
